File: javaBootcamp/src/main/java/model/content/Bootcamp.java
package model.content;

import model.Dev;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class Bootcamp extends Content {

    private final LocalDate initialDate;
    private final LocalDate finalDate;
    private final Set<Dev> devs;
    private final LinkedHashSet<Content> contents;

    public Bootcamp(String title, String description) {
        super(title, description);
        this.initialDate = LocalDate.now();
        this.finalDate = initialDate.plusDays(45);
        this.devs = new HashSet<>();
        this.contents = new LinkedHashSet<>();
    }

    public Bootcamp(String title, String description, Set<Dev> devs, LinkedHashSet<Content> contents) {
        super(title, description);
        this.initialDate = LocalDate.now();
        this.finalDate = initialDate.plusDays(45);
        this.devs = devs;
        this.contents = contents;
        this.devs.forEach(dev -> this.contents.forEach(dev::subscribeContent));
    }

    public void subscribe(Dev dev) {
        devs.add(dev);
        this.contents.forEach(dev::subscribeContent);
        dev.subscribeContent(this);
    }

    public void unsubscribe(Dev dev) {
        devs.remove(dev);
        this.contents.forEach(dev::unsubscribeContent);
    }

    public String subscribedDevs() {
        return devs.stream().map(Dev::toString).collect(Collectors.joining("\n"));
    } 
    
    @Override
    public double xpAwarded() {
        return this.contents.size() * 10 * DEFAULT_XP;
    }

    @Override
    public String toString() {
        return "\n-------------++++++++++++++  B  O  O  T  C  A  M  P  ++++++++++++++-------------" +
                "\n:Title: " + getTitle() +
                "\n:Description: " + getDescription() +
                "\n:Initial Date: " + initialDate +
                "\n:Final Date: " + finalDate +
                "\n:XP Awarded: " + xpAwarded() +
                "\n**** CONTENT ****\n" +
                contents.stream().map(Content::toString).collect(Collectors.joining("\n")) +
                "-------------++++++++++++++  B  O  O  T  C  A  M  P  ++++++++++++++-------------\n"
                ;
    }
}

File: kotlinBootcamp/src/main/kotlin/model/content/Bootcamp.kt
package model.content

import model.Dev
import java.time.LocalDate
import java.util.function.Consumer


class Bootcamp : Content {
    private val initialDate: LocalDate
    private val finalDate: LocalDate
    private val devs: MutableSet<Dev>
    private val contents: LinkedHashSet<Content>

    constructor(title: String, description: String) : super(title, description) {
        initialDate = LocalDate.now()
        finalDate = initialDate.plusDays(45)
        devs = HashSet()
        contents = LinkedHashSet()
    }

    constructor(title: String, description: String, devs: MutableSet<Dev>, contents: LinkedHashSet<Content>)
            : super(title, description) {
        initialDate = LocalDate.now()
        finalDate = initialDate.plusDays(45)
        this.devs = devs
        this.contents = contents
        this.devs.forEach { dev: Dev ->
            this.contents.forEach  { content ->
                dev.subscribeContent(
                    content
                )
            }
        }
    }

    fun subscribe(dev: Dev) {
        devs.add(dev)
        contents.forEach(Consumer { content ->
            dev.subscribeContent(
                content
            )
        })
        dev.subscribeContent(this)
    }

    fun unsubscribe(dev: Dev) {
        devs.remove(dev)
        contents.forEach(Consumer { content ->
            dev.unsubscribeContent(
                content
            )
        })
    }

    fun subscribedDevs(): String {
        return devs.joinToString("\n") { obj: Dev -> obj.toString() }
    }

    override fun xpAwarded(): Double {
        return contents.size * 10 * DEFAULT_XP
    }

    override fun toString(): String {
        return """
            
-------------++++++++++++++  B  O  O  T  C  A  M  P  ++++++++++++++-------------
:Title: $title
:Description: $description
:Initial Date: $initialDate
:Final Date: $finalDate
:XP Awarded: ${xpAwarded()}
**** CONTENT ****
${
contents.joinToString("\n") { obj: Content -> obj.toString() }
}-------------++++++++++++++  B  O  O  T  C  A  M  P  ++++++++++++++-------------

            """
    }
}